README.md
# 查询 openai 余额

## Install

```shell
npm i openai-billing
```

## Usage

```javascript
import { fetchBilling } from "openai-billing"

// for CommonJS
// const { fetchBilling } = require("openai-billing")

const apikey = 'YOUR OPENAI KEY'
const result = await fetchBilling(apikey, { baseURL: 'https://api.openai.com' })

if (result.error) {
    const { code, message } = result.error
    console.error(`Error: ${code} - ${message}`)
}
else {
    const { total, used, remain, expiresTime } = result
    console.log(`
额度总量: 💵 $ ${total}
已用额度: 💵 $ ${used}
剩余额度: 💵 $ ${remain}
过期时间：🪫  ${new Date(expiresTime).toLocaleDateString()}
    `)
}

```

## Use in browser

```html
<script src="https://cdn.jsdelivr.net/npm/openai-billing@2">
```

```javascript
const apikey = 'YOUR OPENAI KEY'
const result = await OpenaiBilling.fetchBilling(apikey, { baseURL: 'https://api.openai.com' })

// ...
```


CHANGELOG.md
# Changelog

## [2.1.0](https://github.com/LinLin00000000/openai-billing/compare/v2.0.0...v2.1.0) (2023-05-03)


### Features

* add support and documentation for browser iife usage ([f4305e0](https://github.com/LinLin00000000/openai-billing/commit/f4305e0fa97fc6d364f9578164a8ee41408556ad))

## [2.0.0](https://github.com/LinLin00000000/openai-billing/compare/v1.1.1...v2.0.0) (2023-04-20)


### ⚠ BREAKING CHANGES

* remove default export for cjs compatibility

### Features

* remove default export for cjs compatibility ([ad6d7ab](https://github.com/LinLin00000000/openai-billing/commit/ad6d7ab8a6d844927d59eb86b5935118833de1d6))

## [1.1.1](https://github.com/LinLin00000000/openai-billing/compare/v1.1.0...v1.1.1) (2023-04-20)


### Bug Fixes

* (node:43476) [DEP0128] DeprecationWarning: Invalid 'main' field in ... ([d326a57](https://github.com/LinLin00000000/openai-billing/commit/d326a5770f2b86b4bd4173d17d69774b180896a9))


index.js
import fetch from 'cross-fetch'

/**
 * @param {string} apikey
 * @param {object} options
 * @param {string} options.baseURL - openai api base url, need http/https prefix, default to 'https://api.openai.com'
 * @returns {Promise<{ total: number, used: number, remain: number, expiresTime: number } | { error: { code: string, message: string } }>}
 */
export async function fetchBilling(
    apikey,
    { baseURL = 'https://api.openai.com' } = {}
) {
    const headers = {
        'Authorization': 'Bearer ' + apikey,
        'Content-Type': 'application/json',
    }

    try {
        const subscription = await fetch(`${baseURL}/dashboard/billing/subscription`, { headers }).then(res => res.json())

        if (subscription.error) {
            return {
                error: {
                    code: subscription.error.code,
                    message: subscription.error.message,
                },
            }
        }

        const now = new Date()
        const expiresTime = subscription.access_until * 1000

        if (now > expiresTime) {
            return {
                error: {
                    code: 'grant_expired',
                    message: 'The grant for the account has expired',
                },
            }
        }

        // 如果用户绑卡，额度每月会刷新，则起始日期为本月第一天，否则为三个月前
        const startDate = subscription.has_payment_method
            ? new Date(now.getFullYear(), now.getMonth())
            : new Date(now - 90 * 24 * 60 * 60 * 1000)

        const endDate = new Date(now.getTime() + 24 * 60 * 60 * 1000)
        const usage = await fetch(`${baseURL}/dashboard/billing/usage?start_date=${formatDate(startDate)}&end_date=${formatDate(endDate)}`, { headers }).then(res => res.json())

        const total = subscription.hard_limit_usd
        const used = usage.total_usage / 100
        const remain = total - used

        return {
            total,
            used,
            remain,
            expiresTime
        }
    } catch (err) {
        return {
            error: {
                code: err.code ?? 'unknown',
                message: err.message ?? err.stack,
            },
        }
    }
}

/**
 * @param {Date} date
 */
function formatDate(date) {
    return date.toISOString().slice(0, 10)
}


index.bs.js
// Generated by ReScript, PLEASE EDIT WITH CARE


import fetch from 'cross-fetch'

/**
 * @param {string} apikey
 * @param {object} options
 * @param {string} options.baseURL - openai api base url, need http/https prefix, default to 'https://api.openai.com'
 * @returns {Promise<{ total: number, used: number, remain: number, expiresTime: number } | { error: { code: string, message: string } }>}
 */
export async function fetchBilling(
    apikey,
    { baseURL = 'https://api.openai.com' } = {}
) {
    const headers = {
        'Authorization': 'Bearer ' + apikey,
        'Content-Type': 'application/json',
    }

    try {
        const subscription = await fetch(`${ baseURL }/dashboard/billing/subscription`, { headers }).then(res => res.json())

        if (subscription.error) {
            return {
                error: {
                    code: subscription.error.code,
                    message: subscription.error.message,
                },
            }
        }

        const now = new Date()
        const expiresTime = subscription.access_until * 1000

        if (now > expiresTime) {
            return {
                error: {
                    code: 'grant_expired',
                    message: 'The grant for the account has expired',
                },
            }
        }

        // 如果用户绑卡，额度每月会刷新，则起始日期为本月第一天，否则为三个月前
        const startDate = subscription.has_payment_method
            ? new Date(now.getFullYear(), now.getMonth())
            : new Date(now - 90 * 24 * 60 * 60 * 1000)

        const endDate = new Date(now.getTime() + 24 * 60 * 60 * 1000)
        const usage = await fetch(`${ baseURL }/dashboard/billing/usage?start_date=${ formatDate(startDate) }&end_date=${ formatDate(endDate) }`, { headers }).then(res => res.json())

        const total = subscription.hard_limit_usd
        const used = usage.total_usage / 100
        const remain = total - used

        return {
            total,
            used,
            remain,
            expiresTime
        }
    } catch (err) {
        return {
            error: {
                code: err.code ? err.code : 'unknown',
                message: err.message ? err.message : err.stack,
            },
        }
    }
}
;

function formatDate(date) {
  return date.toISOString().slice(0, 10);
}

export {
  formatDate ,
}
/*  Not a pure module */
